src/Compilers/CSharp/Test/Semantic/Semantics/SpanStackSafetyTests.cs

```diff
@@ -1684,5 +1684,123 @@ static void M2(__arglist)
                 //         M2(__arglist(ref s2, ref s1));  // two
                 Diagnostic(ErrorCode.ERR_CallArgMixing, "M2(__arglist(ref s2, ref s1))").WithArguments("C.M2(__arglist)", "__arglist").WithLocation(17, 9));
         }
+
+        [Fact]
+        [WorkItem(44588, "https://github.com/dotnet/roslyn/issues/44588")]
+        public void SwitchDefaultLocalInitialization_01()
+        {
+            var source = @"
+using System;
+
+struct Struct2
+{
+    public static ReadOnlySpan<T> GetSpan1<T>(int x)
+    {
+        return x switch
+        {
+            0 => default,
+            _ => default,
+        };
+    }
+
+    public static ReadOnlySpan<T> GetSpan2<T>(int x)
+    {
+        ReadOnlySpan<T> span;
+
+        span = x switch
+        {
+            0 => default,
+            _ => default,
+        };
+
+        return span;
+    }
+
+    public static ReadOnlySpan<T> GetSpan3<T>(int x)
+    {
+        ReadOnlySpan<T> span = x switch
+        {
+            0 => default,
+            _ => default,
+        };
+
+        return span;
+    }
+
+    public static ReadOnlySpan<T> GetSpan4<T>(int x)
+    {
+        ReadOnlySpan<T> span = default;
+
+        return span;
+    }
+}";
+            CreateCompilationWithMscorlibAndSpan(source).VerifyDiagnostics(
+                );
+        }
+
+        [Fact]
+        [WorkItem(44588, "https://github.com/dotnet/roslyn/issues/44588")]
+        public void SwitchDefaultLocalInitialization_02()
+        {
+            var source = @"
+using System;
+
+struct Struct2
+{
+    public static Span<byte> GetSpan1(int x)
+    {
+        return x switch
+        {
+            0 => stackalloc byte[10], // 1
+            _ => default,
+        };
+    }
+
+    public static Span<byte> GetSpan2(int x)
+    {
+        Span<byte> span;
+
+        span = x switch
+        {
+            0 => stackalloc byte[10], // 2
+            _ => default,
+        };
+
+        return span;
+    }
+
+    public static Span<byte> GetSpan3(int x)
+    {
+        Span<byte> span = x switch
+        {
+            0 => stackalloc byte[10],
+            _ => default,
+        };
+
+        return span; // 3
```

Comment:
>span; // 3 [](start = 15, length = 10)

Why is this error in `GetSpan3` reported on the `return` while the error in `GetSpan2` is reported within the `switch` expression? I'm curious in case the difference points to an issue.

Reply:
There is no problem in `GetSpan3` until the `return`. The switch expression simply computes a `Span` value that cannot be returned, and is used to declare the local `span` which therefore cannot be returned. It is only when you attempt to return that value that there is an issue.

---
In reply to: [441674063](https://github.com/dotnet/roslyn/pull/45242#discussion_r441674063) [](ancestors = 441674063)

Reply:
> There is no problem in `GetSpan3` until the `return`.

Isn't that true for `GetSpan2` as well? It looks like the only difference between the two cases is whether the local is assigned in the declaration.

Reply:
We don't track whether a variable can be returned based on the flow of data (like we do with nullable) but based on the "returnable" property of the local computed at the declaration point.
In `GetSpan2` the local is returnable so we cannot even assign the switch expression to it.
In `GetSpan3` the local is not returnable (because its initializing expression isn't returnable) so the assignment of the switch expression to it isn't a problem.

---
In reply to: [441738383](https://github.com/dotnet/roslyn/pull/45242#discussion_r441738383) [](ancestors = 441738383)